### ISSUE TITLE
NUT communication error: configuration unreadable

### ISSUE THREAD

**Issue by peterdenhaan** [rockstor/rockstor-core#1073]
On a fresh install of Rockstor (3.8-9.02) I had problems getting UPS support to work until I fixed group ownership in `/etc/ups/`. I cannot find an issue for this here in Git, so I'm opening one.

I am using an USB connected APC Smart-UPS 750 with the `usbhid-ups` driver. Looking through the journal, `upsdrvctl` started the driver just fine, but `upsd` was not initialising properly:
```
Dec 29 22:22:59 animalfarm systemd[1]: Starting Network UPS Tools - power device driver controller...
Dec 29 22:22:59 animalfarm systemd-tmpfiles[5344]: Failed to open '/etc/tmpfiles.d/nut-run.conf', ignoring: No such file or directory
Dec 29 22:22:59 animalfarm upsdrvctl[5347]: Duplicate driver instance detected! Terminating other driver!
Dec 29 22:22:59 animalfarm usbhid-ups[5092]: Signal 15: exiting
Dec 29 22:23:05 animalfarm upsdrvctl[5347]: interrupt pipe disabled (add 'pollonly' flag to 'ups.conf' to get rid of this message)
Dec 29 22:23:05 animalfarm upsdrvctl[5347]: Using subdriver: APC HID 0.95
Dec 29 22:23:05 animalfarm upsdrvctl[5347]: Network UPS Tools - Generic HID driver 0.38 (2.7.2)
Dec 29 22:23:05 animalfarm upsdrvctl[5347]: USB communication driver 0.32
Dec 29 22:23:05 animalfarm upsdrvctl[5347]: Network UPS Tools - UPS driver controller 2.7.2
Dec 29 22:23:05 animalfarm systemd[1]: Started Network UPS Tools - power device driver controller.
Dec 29 22:23:05 animalfarm systemd[1]: Starting Network UPS Tools - power devices information server...
Dec 29 22:23:05 animalfarm systemd[1]: Started Network UPS Tools - power devices information server.
Dec 29 22:23:05 animalfarm usbhid-ups[5371]: Startup successful
Dec 29 22:23:05 animalfarm upsd[5372]: fopen /var/run/nut/upsd.pid: No such file or directory
Dec 29 22:23:05 animalfarm upsd[5372]: listening on 0.0.0.0 port 3493
Dec 29 22:23:05 animalfarm upsd[5372]: listening on 0.0.0.0 port 3493
Dec 29 22:23:05 animalfarm upsd[5372]: Can't open /etc/ups/ups.conf: Can't open /etc/ups/ups.conf: Permission denied
Dec 29 22:23:05 animalfarm upsd[5372]: Can't open /etc/ups/ups.conf: Can't open /etc/ups/ups.conf: Permission denied
Dec 29 22:23:05 animalfarm upsd[5372]: Network UPS Tools upsd 2.7.2
```
The journal subsequently filled up with connection failure messages every 15 seconds (retries):
```
Dec 29 22:26:03 animalfarm upsmon[6067]: UPS: ups@localhost (master) (power value 1)
Dec 29 22:26:03 animalfarm upsmon[6067]: UPS [ups@localhost]: connect failed: Connection failure: Connection refused
Dec 29 22:26:03 animalfarm upsmon[6067]: Communications with UPS ups@localhost lost
```
The clue is in the first journal section above: `/etc/ups/ups.conf` cannot be opened by `upsd`. That looks like a problem. I tried to figure out why:
```
-rw-r-----  1 root root  1616 Dec 29 22:19 nut.conf
-rw-r-----  1 root root  3784 Dec 29 22:19 ups.conf
-rw-r-----  1 root root  4655 Dec 29 22:19 upsd.conf
-rw-r-----  1 root root  2235 Dec 29 22:19 upsd.users
-rw-r-----  1 root root 15708 Dec 29 22:19 upsmon.conf
-rw-r-----. 1 root nut   4631 Dec 29 22:19 upssched.conf
```
Unless `upsd` is run as root - I hope not - it will indeed be unable to read its configuration file. On a hunch, I changed group ownership of all the files above to `nut`. The UPS code now is happy as far as I can see:
```
Dec 29 22:27:08 animalfarm upsdrvctl[6315]: Network UPS Tools - UPS driver controller 2.7.2
Dec 29 22:27:08 animalfarm systemd[1]: Started Network UPS Tools - power device driver controller.
Dec 29 22:27:08 animalfarm systemd[1]: Starting Network UPS Tools - power devices information server...
Dec 29 22:27:08 animalfarm systemd[1]: Started Network UPS Tools - power devices information server.
Dec 29 22:27:08 animalfarm usbhid-ups[6317]: Startup successful
Dec 29 22:27:08 animalfarm upsd[6318]: listening on 0.0.0.0 port 3493
Dec 29 22:27:08 animalfarm upsd[6318]: Connected to UPS [ups]: usbhid-ups-ups
Dec 29 22:27:08 animalfarm upsd[6318]: fopen /var/run/nut/upsd.pid: No such file or directory
Dec 29 22:27:08 animalfarm upsd[6318]: listening on 0.0.0.0 port 3493
Dec 29 22:27:08 animalfarm upsd[6318]: Connected to UPS [ups]: usbhid-ups-ups
Dec 29 22:27:08 animalfarm systemd[1]: Reloading.
Dec 29 22:27:08 animalfarm systemd[1]: Starting Network UPS Tools - power device monitor and shutdown controller...
Dec 29 22:27:08 animalfarm systemd-tmpfiles[6329]: Failed to open '/etc/tmpfiles.d/nut-run.conf', ignoring: No such file or directory
Dec 29 22:27:08 animalfarm systemd[1]: Started Network UPS Tools - power device monitor and shutdown controller.
Dec 29 22:27:08 animalfarm upsmon[6331]: fopen /var/run/nut/upsmon.pid: No such file or directory
Dec 29 22:27:08 animalfarm upsmon[6331]: Using power down flag file /etc/ups/killpower
Dec 29 22:27:08 animalfarm upsmon[6331]: UPS: ups@localhost (master) (power value 1)
Dec 29 22:27:08 animalfarm upsd[6318]: User monuser@127.0.0.1 logged into UPS [ups]
Dec 29 22:27:08 animalfarm upsd[6318]: User monuser@127.0.0.1 logged into UPS [ups]
```
Given that this was a clean Rockstor install, I suspect that group ownership is not correctly set when these configuration files are first installed or generated.

Thanks for reading, and thanks for the amazing work you guys are doing on Rockstor.

**Comment by phillxnet** [comment]
@peterdenhaan Thanks for the detailed report on this, I will look into this hopefully soon. I do remember having to adjust file access rights so the nut user could read them and did so in the code when the config is applied (from my rather faint memory) but maybe things have changed. Only the parts that need to run as root are run as such and we pretty much go with the defaults of nut for permissions then edit the files programatically. Since you are obviously technically aware you might like to peruse the wiki on how nut is implemented in Rockstor via the wiki entry, they are like the technical manuals for how stuff is done. http://forum.rockstor.com/t/network-ups-tools-nut-rockstors-implementation/553
Please feel free to comment if you see anything amis.
I am due to look again at nut as part of https://github.com/rockstor/rockstor-core/issues/982
When I last tested the nut system I did so with a serial and usb and usb-to-serial adapter and all was well so not sure what's happened there really.

Thanks again for pointing this out and for sharing you fix. I will try to get to this soon.

Glad you got it working.

**Comment by phillxnet** [comment]
@peterdenhaan I have just had a quick look at this and when I configure NUT-UPS via the System - Services spanner next to the name as per the docs:-
http://rockstor.com/docs/ups-setup/ups_setup.html
my config files are as follows:-
```
ls -la /etc/ups/
total 44
drwxr-xr-x. 1 root root   118 Dec 21 13:46 .
drwxr-xr-x. 1 root root  3600 Jan 12 10:34 ..
-rw-r-----. 1 root nut   1616 Jan 12 10:54 nut.conf
-rw-r-----. 1 root nut   3783 Jan 12 10:54 ups.conf
-rw-r-----. 1 root nut   4657 Jan 12 10:54 upsd.conf
-rw-r-----. 1 root nut   2238 Jan 12 10:54 upsd.users
-rw-r-----. 1 root nut  15715 Jan 12 10:54 upsmon.conf
-rw-r-----. 1 root nut   4631 Jan 12 10:54 upssched.conf
```
Note that they are already group nut readable. The service starts and reports on the state of the UPS as expected (ie via upsc ups@nutserver.lan for a netclient install).
This was with a 3.8-10.15 testing channel updates Rockstor but nothing has changed in nut since your install version, code wise at least. I will have a look at the clean install from 3.8-10 to double check but I do remember having to address this issue when the nut / ups feature was added in pull request #892 . Note line 201 in /opt/rockstor/src/rockstor/system/nut.py in the config_nut method / procedure. However this only does a chmod not a chown but the above directory listing does show nut as group owner in my example.

I will look into this a little more and report back here.

**Comment by phillxnet** [comment]
OK, I have just done a fresh install of 3.8-9.02 (from the iso) and have the same outcome nut config file wise; not that this is before any updates:-
```
ls -la /etc/ups
total 40
drwxr-xr-x. 1 root root   118 Jan 12 11:48 .
drwxr-xr-x. 1 root root  3526 Jan 12 11:52 ..
-rw-r-----. 1 root nut   1538 Feb 23  2015 nut.conf
-rw-r-----. 1 root nut   3668 Feb 23  2015 ups.conf
-rw-r-----. 1 root nut   4578 Feb 23  2015 upsd.conf
-rw-r-----. 1 root nut   2131 Feb 23  2015 upsd.users
-rw-r-----. 1 root nut  15178 Feb 23  2015 upsmon.conf
-rw-r-----. 1 root nut   3891 Feb 23  2015 upssched.conf
```
And then after updating using the testing channel to 3.8-10.16 these permissions remained unchanged.  However after doing an identical config to my last post I have reproduced your nut config file ownership ie:-
```
ls -la /etc/ups/
total 44
drwxr-xr-x. 1 root root   118 Jan 12 12:21 .
drwxr-xr-x. 1 root root  3542 Jan 12 12:13 ..
-rw-r-----  1 root root  1616 Jan 12 12:21 nut.conf
-rw-r-----  1 root root  3783 Jan 12 12:21 ups.conf
-rw-r-----  1 root root  4657 Jan 12 12:21 upsd.conf
-rw-r-----  1 root root  2238 Jan 12 12:21 upsd.users
-rw-r-----  1 root root 15715 Jan 12 12:21 upsmon.conf
-rw-r-----. 1 root nut   4631 Jan 12 12:21 upssched.conf
```
Thanks for bringing my attention to this matter; my suspicion is that background updates have changed the default mask as this didn't use to happen.
I will attempt a fix shorty and use this issue as a reference.

**Comment by phillxnet** [comment]
With the last patch an existing system with non working nut due to nut config file access rights changes (as in @peterdenhaan example) had it's config file access rights changed from the problematic:-
```ls -la /etc/ups/
total 44
drwxr-xr-x. 1 root root   118 Jan 12 14:44 .
drwxr-xr-x. 1 root root  3542 Jan 12 14:49 ..
-rw-r-----  1 root root  1617 Jan 12 14:44 nut.conf
-rw-r-----  1 root root  3811 Jan 12 14:44 ups.conf
-rw-r-----  1 root root  4657 Jan 12 14:44 upsd.conf
-rw-r-----  1 root root  2239 Jan 12 14:44 upsd.users
-rw-r-----  1 root root 15712 Jan 12 14:44 upsmon.conf
-rw-r-----. 1 root nut   4631 Jan 12 14:44 upssched.conf
```
to the following:-
```
ls -la /etc/ups/
total 44
drwxr-xr-x. 1 root root   118 Jan 12 14:49 .
drwxr-xr-x. 1 root root  3542 Jan 12 14:49 ..
-rw-r-----  1 root nut   1617 Jan 12 14:49 nut.conf
-rw-r-----  1 root nut   3811 Jan 12 14:49 ups.conf
-rw-r-----  1 root nut   4657 Jan 12 14:49 upsd.conf
-rw-r-----  1 root nut   2239 Jan 12 14:49 upsd.users
-rw-r-----  1 root nut  15712 Jan 12 14:49 upsmon.conf
-rw-r-----. 1 root nut   4631 Jan 12 14:49 upssched.conf
```
N.B. this takes effect after a successful nut re-config via the GUI. In this case the nut config then functioned as expected.

**Comment by peterdenhaan** [comment]
Fantastic. Thanks @phillxnet.

**Comment by phillxnet** [comment]
@peterdenhaan Cheers, note though that although I think this issue is fixed the linked issue #1094 is still outstanding but I am chipping away at it.

**Comment by phillxnet** [comment]
@peterdenhaan OK, I think I've sorted the nut config screen in the submitted pull request as well so if it passes review then in time this should be included in subsequent updates.

Thanks again for reporting this; turns out to be quite a corner case as I had difficulty reproducing it and in may instances the config files kept their root.nut ownership just dandy. Was about to report I couldn't reproduce when all of a sudden it occurred. I was quite relieved as then at least I could test the proposed fix.
